samcli/commands/init/init_templates.py

```diff
@@ -171,3 +173,46 @@ def is_dynamic_schemas_template(self, package_type, app_template, runtime, base_
             if option.get("appTemplate") == app_template:
                 return option.get("isDynamicTemplate", False)
         return False
+
+    def get_preprocessed_manifest(self):
+        """
+        This method get the manifest cloned from the git repo and preprocessed it.
+
+        Raises:
+            InvalidInitTemplateError: Exception raise when template is invalid or missing a parameter
+
+        Returns:
+            dict: This is preprocessed manifest with the use_case as key
+        """
+        manifest_dump = requests.get(self._manifest_url)
+        if not manifest_dump:
+            raise InvalidInitTemplateError(
+                "Can't retrieve templates manifest from {url}".format(url=self._manifest_url)
+            )
+        preprocessed_manifest = {}
+        manifest_body = manifest_dump.json()
+
+        for template_runtime in manifest_body:
+            template_list = manifest_body[template_runtime]
+            for template in template_list:
+                if "useCaseName" not in template:
+                    raise InvalidInitTemplateError(
+                        "Template {name} missing use case descriptor".format(name=template["displayName"])
+                    )
+                use_case = template["useCaseName"]
+                if use_case not in preprocessed_manifest:
```

Comment:
Instead of multiple if statements could we adopt a matcher pattern, so that we can supply a list of matchers (eg: predicates) that would short circuit the error process, makes the code easier to read. Agree on the fact the sam init codebase has some tech debt, we can fix it a little bit at a time. :)

Reply:
I don't understand what you mean by `matcher pattern` but I have extracted the `if statements` into a different method to handle validation of the properties.

Reply:
Think of the validations you are performing to be of thier objects, that way you could do something like

```
for validator in validators:
      validator.validate(template)
```

Reply:
I think you may simply be able to do a:
```
use_case = preprocessed_manifest.get(use_case_name, {})
package_types = use_case.get(runtime, {}).get(package_type, [])
package_types.append(template)
preprocessed_manifest[use_case_name] = use_case
```

You may also want to rename `validate_value_in_template` to `get_template_value` and have it return the value directly